File: leetcode/1346-check-if-n-and-its-double-exist/CheckIfNAndItsDoubleExist.java
class Solution {
    public boolean checkIfExist(int[] a) {
        int[] cache = new int[2000 * 2 + 1];

        for (int i = 0; i < a.length; i++) {
            // int currentValue = a[i];

            if (cache[a[i] * 2 + 2000] != 0 || (a[i] % 2 == 0 && cache[a[i] / 2 + 2000] != 0)) {
                return true;
            }

            cache[a[i] + 2000] = i + 1;
        }

        return false;
    }
}

public class CheckIfNAndItsDoubleExist {
    public static void main(String[] arguments) {
        // int[] a = { 0, 0 };
        // int[] a = { 10, 2, -5, 3 };
        int[] a = { -1, 2, 0, -4 };

        System.out.println((new Solution()).checkIfExist(a));
    }
}

File: leetcode/1001-2000/1301-1400/1346-check-if-n-and-its-double-exist/CheckIfNAndItsDoubleExist.ts
function checkIfExist(numbers: number[]): boolean {
  const cache = {}

  for (const number of numbers) {
    if (cache[number * 2] !== undefined || (number % 2 === 0 && cache[number / 2] !== undefined)) {
      return true;
    }

    cache[number] = true;
  }

  return false;
};

console.log(checkIfExist([10, 2, 5, 3]));
console.log(checkIfExist([3, 1, 7, 11]));
